virtcontainers/kata_agent.go

```diff
@@ -843,6 +848,43 @@ func (k *kataAgent) handleBlockVolumes(c *Container) []*grpc.Storage {
 	return volumeStorages
 }
 
+func (k *kataAgent) handlePidNamespace(grpcSpec *grpc.Spec, sandbox *Sandbox) (bool, error) {
+	sharedPidNs := false
+
+	for i, ns := range grpcSpec.Linux.Namespaces {
+		if ns.Type == string(specs.PIDNamespace) {
+			if ns.Path != "" {
+				if sandbox.state.Pid > 0 {
+					pidNsPath := fmt.Sprintf("/proc/%d/ns/pid", sandbox.state.Pid)
+
+					//  Check if pid namespace path is the same as the sandbox
+					if ns.Path == pidNsPath {
+						sharedPidNs = true
+					} else {
+						// Check for symlinks
+						ln, err := filepath.EvalSymlinks(ns.Path)
+						if err != nil {
+							return sharedPidNs, err
+						}
+
+						// We have arbitrary pid namespace path here
+						if ln != pidNsPath {
+							return sharedPidNs, fmt.Errorf("Pid namespace path other than sandbox")
+						}
+						sharedPidNs = true
+					}
+				}
+			}
+
+			// Remove pid namespace.
+			grpcSpec.Linux.Namespaces = append(grpcSpec.Linux.Namespaces[:i], grpcSpec.Linux.Namespaces[i+1:]...)
```

Comment:
I think it would be more appropriate to handle the spec modification from `constraintGRPCSpec(grpcSpec)`, by providing `sharedPidNs` flag. This way, based on the value of `sharedPidNs`, the spec will be updated accordingly. Otherwise, we're modifying the spec in too many places IMO.

Reply:
The reason I did it this way was to keep all the logic related to PID namespaces in a single function. If I move it to `constraintGRPCSpec` , it still means that we have to loop over all the namespaces yet again.

Reply:
Yes good point, then if we go with this solution, you should add some comments about the function itself, describing the fact that it modifies the grpcSpec structure.